### PR TITLE
Feature: Microsoft Actions integration

### PR DESCRIPTION
**Pull request by vg-svitla**

### CODE REVIEW

File: Microsoft/action_change_user_password.json [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        "type": "string"
+      },
+      "username": {
+        "description": "Username to do authentication on remote server",
+        "type": "string"
+      },
+      "password": {
+        "description": "Password to do authentication on remote server",
+        "type": "string"
+      },
```
Comment (squioc) [inline Microsoft/action_change_user_password.json:16]: I think that these parameters should be declared as module parameters in the manifest.json

File: Microsoft/action_change_user_password.json [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      "new_password": {
+        "description": "New password for user",
+        "type": "string"
+      }
+    },
+    "required": [
+      "server",
+      "username",
+      "password",
```
Comment (squioc) [inline Microsoft/action_change_user_password.json:29]: same as above

File: Microsoft/action_change_user_password.json [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      "user_to_update",
+      "new_password"
+    ],
+    "secrets": [
+      "password",
```
Comment (squioc) [inline Microsoft/action_change_user_password.json:34]: Same as above

File: Microsoft/action_disable_users.json [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        "type": "string"
+      },
+      "username": {
+        "description": "Username to do authentication on remote server",
+        "type": "string"
+      },
+      "password": {
+        "description": "Password to do authentication on remote server",
+        "type": "string"
+      },
```
Comment (squioc) [inline Microsoft/action_disable_users.json:16]: same as above

File: Microsoft/action_enable_users.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        "type": "string"
+      },
+      "username": {
+        "description": "Username to do authentication on remote server",
+        "type": "string"
+      },
+      "password": {
+        "description": "Password to do authentication on remote server",
+        "type": "string"
+      },
```
Comment (squioc) [inline Microsoft/action_enable_users.json:16]: Same

File: Microsoft/action_enable_users.json [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      }
+    },
+    "required": [
+      "server",
+      "username",
+      "password"
+    ],
+    "secrets": [
+      "password"
+    ],
```
Comment (squioc) [inline Microsoft/action_enable_users.json:39]: Same

File: Microsoft/action_disable_users.json [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      }
+    },
+    "required": [
+      "server",
+      "username",
+      "password"
+    ],
+    "secrets": [
+      "password"
+    ],
```
Comment (squioc) [inline Microsoft/action_disable_users.json:39]: Same

File: Microsoft/actions/change_user_password_action.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+from client.commands import PowershellCommand
+from client.windows_client import WindowsRemoteClient
+
+
+class ChangeUserPasswordActionConfig(BaseModel):
+    """Config for ChangeUserPasswordAction."""
+
+    username: str
+    password: str
+    server: str
```
Comment (squioc) [inline Microsoft/actions/change_user_password_action.py:16]: To move to the Module Configuration

File: Microsoft/actions/change_user_password_action.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        Args:
+            arguments: dict[str, Any]
+
+        Returns:
+            dict:
+        """
+        config = ChangeUserPasswordActionConfig(**arguments)
+        command = PowershellCommand.change_user_password(config.user_to_update, config.new_password)
+
+        WindowsRemoteClient(config.server, config.username, config.password).execute_command(command)
```
Comment (squioc) [inline Microsoft/actions/change_user_password_action.py:37]: With the move of these parameters, I suggest this change
```suggestion
        WindowsRemoteClient(self.module.configuration.server, self.module.configuration.username, self.module.configuration.password).execute_command(command)
```

File: Microsoft/actions/disable_users_action.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+from client.commands import PowershellCommand
+from client.windows_client import WindowsRemoteClient
+
+
+class DisableUsersActionConfig(BaseModel):
+    """Config for DisableUsersAction."""
+
+    username: str
+    password: str
+    server: str
```
Comment (squioc) [inline Microsoft/actions/disable_users_action.py:16]: Same

File: Microsoft/actions/disable_users_action.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        Args:
+            arguments: dict[str, Any]
+
+        Returns:
+            dict:
+        """
+        config = DisableUsersActionConfig(**arguments)
+        command = PowershellCommand.disable_users(usernames=config.users, sids=config.sids)
+
+        WindowsRemoteClient(config.server, config.username, config.password).execute_command(command)
```
Comment (squioc) [inline Microsoft/actions/disable_users_action.py:37]: Same

File: Microsoft/actions/enable_users_action.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+from client.commands import PowershellCommand
+from client.windows_client import WindowsRemoteClient
+
+
+class EnableUsersActionConfig(BaseModel):
+    """Config for EnableUsersAction."""
+
+    username: str
+    password: str
+    server: str
```
Comment (squioc) [inline Microsoft/actions/enable_users_action.py:16]: Same

File: Microsoft/actions/enable_users_action.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        Args:
+            arguments: dict[str, Any]
+
+        Returns:
+            dict:
+        """
+        config = EnableUsersActionConfig(**arguments)
+        command = PowershellCommand.enable_users(usernames=config.users, sids=config.sids)
+
+        WindowsRemoteClient(config.server, config.username, config.password).execute_command(command)
```
Comment (squioc) [inline Microsoft/actions/enable_users_action.py:37]: Same

File: Microsoft/pyproject.toml [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+name = "Sekoia.io Automation Module to work with Microsoft Server Actions events"
+version = "0.0"
+description = ""
+authors = []
+
+[tool.poetry.dependencies]
+python = ">=3.10,<3.12"
+orjson = "^3.7.7"
+python-dateutil = "^2.8.2"
+sekoia-automation-sdk = { version = "^1.8.1", extras = ["all"] }
```
Comment (squioc) [inline Microsoft/pyproject.toml:11]: Can you update the sdk to the last version?

File: Microsoft/pyproject.toml [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+async-lru = "^2.0.2"
+aiohttp = "^3.8.5"
+aiolimiter = "^1.1.0"
```
Comment (squioc) [inline Microsoft/pyproject.toml:14]: Are these dependencies useful?

File: Microsoft/pyproject.toml [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+pytest-asyncio = "*"
+pytest-xdist = "*"
+pytest-env = "*"
+pytest-flake8 = "*"
+Faker = "^13.12.0"
+autoflake = "^1.4"
+isort = "^5.10.1"
+black = "^22.3.0"
+mypy = "^0.960"
+aioresponses = "^0.7.4"
```
Comment (squioc) [inline Microsoft/pyproject.toml:29]: Same

File: Microsoft/pyproject.toml [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+sekoia-automation-sdk = { version = "^1.8.1", extras = ["all"] }
+async-lru = "^2.0.2"
+aiohttp = "^3.8.5"
+aiolimiter = "^1.1.0"
+pywinrm = {extras = ["credssp"], version = "^0.4.3"}
+
+[tool.poetry.dev-dependencies]
+pytest = "*"
+pytest-cov = "*"
+pytest-asyncio = "*"
```
Comment (squioc) [inline Microsoft/pyproject.toml:20]: Same

File: Microsoft/client/commands.py [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+    @staticmethod
+    def get_local_users() -> "PowershellCommand":
+        """
+        Get local users.
+
+        Returns:
+            PowershellCommand:
+        """
+        return PowershellCommand(command="Get-LocalUser")
```
Comment (squioc) [inline Microsoft/client/commands.py:78]: Is used? Can we create an action for it? or remove it if not used?

**Review by squioc — CHANGES_REQUESTED** [review]
Thanks for the job.
Some suggestions and changes to apply.

File: Microsoft/client/commands.py (hunk 18 above)
Comment (vg-svitla) [reply Microsoft/client/commands.py:78]: No it is test command. Removed. Thanks!

File: Microsoft/pyproject.toml (hunk 17 above)
Comment (vg-svitla) [reply Microsoft/pyproject.toml:20]: By default I use `@pytest.mark.asyncio` in tests to speed them up a bit

**Review by squioc — APPROVED** [review]
Nice. I wonder if the user list and the sid list have to be exclusive (ie, If we provide a list of users and a list of sids to the actions, the list of sids won't be processed)? If you keep this behavior, we have to be clear if the description to avoid misusage from the customer.

**Pull request merged**